pennylane/visualize/visualize.py

```diff
@@ -0,0 +1,351 @@
+        """
+
+        return (self._x_log, self._y_log)
+
+    @property
+    def param_data(self):
+        """
+        Returns a tuple of the recorded optimization steps, and the corresponding variational parameters.
+
+        Returns:
+            tuple: a tuple containing lists of optimization steps and parameters values
+        """
+
+        return (self._x_log, self._param_log)
+
+    def __enter__(self):
+        print("Beginning Optimization")
+        print("--------------------------")
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        print("Optimization Complete")
+
+    def update(self, params=None):
+        """
+        Updates the data in the Visualize context manager instance. This data can then be called by the user, or
+        utilized by the active context manager to create text or graphical outputs.
+
+        Keyword Args:
+            params: Variational parameters to be fed into the cost function.
+
+        .. UsageDetails::
+
+            Consider the simple variational quantum circuit and the cost function:
+
+            .. code-block:: python3
+
+                import pennylane as qml
+                from pennylane.visualize import Visualize
+
+                dev = qml.device('default.qubit', wires=3)
+
+                @qml.qnode(dev)
+                def circuit(gamma):
+
+                    qml.RX(gamma, wires=0)
+                    qml.RY(gamma, wires=1)
+                    qml.RX(gamma, wires=2)
+                    qml.CNOT(wires=[0, 2])
+                    qml.RY(gamma, wires=1)
+
+                    return [qml.expval(qml.PauliZ(i)) for i in range(3)]
+
+                def cost(gamma):
+                    return sum(circuit(gamma[0]))
+
+            To utilize the PennyLane visualizations functionality, we wrap the optimization loop
+            in the context manager, and call the ``update()`` method with each optimization step:
+
+            .. code-block:: python3
+
+                optimizer = qml.GradientDescentOptimizer()
+                steps = 5
+                params = np.array([1.])
+
+                with Visualize(steps, cost_fn=cost) as viz:
+                    for i in range(steps):
+                        params = optimizer.step(cost, params)
+                        viz.update(params=params)
+
+            We can then call the values of the cost function for each step:
+
+            >>> steps, cost = viz.cost_data
+            >>> print(cost)
+            [0.29001330299410394, 0.16958978666464564, 0.0554891459073546, -0.05179578982297689, -0.151952780085496]
+        """
+
+        self._step_log += 1
+
+        if self._step_log % self.step_size == 0:
+
+            self._x_log.append(self._step_log)
+
+            if params is not None:
+                self._param_log.append(params)
+                if self.cost_fn is not None:
+                    val = self.cost_fn(params)
```

Comment:
@Lucaman99 note that there is now an `optimizer.step_and_cost()` function, which returns the parameter update _and_ the cost function value in a single quantum evaluation. This will avoid doubling the number of quantum evaluations occurring.